### ISSUE TITLE
Separate ARM translator from IR and code generation

### ISSUE THREAD

**Issue by gdkchan** [Ryujinx/Ryujinx#1125]
Currently the CPU emulator can be thought in two parts:
- ARM code -> IR translation: That's the code inside the Instructions folder, it is used to translate the ARM code to IR that we can later optimize and turn into x86 code.
- IR optimization and X86 backend: That's what actually generates the machine code.

I think they should be separated into 2 projects. First because having a IR -> x86 compiler might be useful for a lot of different things. We can use it to easily generate code at runtime, implement JIT for other CPUs if we want etc. Second it makes testing easier, currently we have tests for the ARM instructions, but we would also benefit from tests to individual IR instructions.

**Comment by Thaodan** [comment]
Would this prepare for ARM support e.g. running the emulator on a mobile Linux OS?

**Comment by gdkchan** [comment]
No, it would allow the JIT to be used by other parts of the project that also need runtime code generation (such as GPU macro emulation, however this already uses .NET JIT, it could be useful in the future if we decide to use a AOT approach where .NET JIT is not available).

**Comment by FICTURE7** [comment]
I am interested into doing this, unless someone else is already working it. I think `Compiler`, `CompilerOptions`, `CompilerContext` can be moved to `ARMeilleure.CodeGen` and `EmitterContext`, `ControlFlowGraph` can be moved to `ARMeilleure.IntermediateRepresentation`. Some analysis can be moved there as well, like SSA construction/deconstruction perhaps under a new namespace `ARMeilleure.CodeGen.Analysis`. There is also PTC, I guess it should be included with the backend?

**Comment by AcK77** [comment]
@FICTURE7 Could you contact me on discord please 😊?

**Comment by FICTURE7** [comment]
I do not have a discord at the moment, maybe I should make one some of these heh. You can reach me through email (whenever I check them <_< atleast).

**Comment by gdkchan** [comment]
> 
> 
> I am interested into doing this, unless someone else is already working it. I think `Compiler`, `CompilerOptions`, `CompilerContext` can be moved to `ARMeilleure.CodeGen` and `EmitterContext`, `ControlFlowGraph` can be moved to `ARMeilleure.IntermediateRepresentation`. Some analysis can be moved there as well, like SSA construction/deconstruction perhaps under a new namespace `ARMeilleure.CodeGen.Analysis`. There is also PTC, I guess it should be included with the backend?

Theres no one working on it right now (to my knowledge). I think in general, everything that is not directly related to Arm should be moved to the new project. The main weird thing right now from what I recall are the `LoadFromContext` and `StoreToContext` IR instructions, removed by the `RegisterUsage` pass. It makes assumptions about how things are currently handled for Arm translation. I also added `ArmEmitterContext` and `EmitterContext` in preparation for this change, so `EmitterContext` is the one that would be on the new JIT project.

I agree with the suggested namespace changes. About PTC, I think it should be part of the Arm recompiler project. I don't think we'll use PTC anywhere else. The backend only needs to know that the constants are relocatable, and return the offsets to them in code.

Anyway, I think most of the complications will arise after the code is moved. I'm happy to give suggestion once you get to that point, it's pretty hard

> 
> 
> @FICTURE7 Could you contact me on discord please 😊?

I think it's fine communicating here. I'm not a fan of Discord myself, the only reason I'm still using it is due to this project.
